src/Orleans/Runtime/GrainInterfaceMap.cs

```diff
@@ -127,49 +128,68 @@ internal void AddMap(GrainInterfaceMap map)
-                    var interfaceTypeKey = GetTypeKey(iface, isGenericGrainClass);
-                    typeToInterfaceData[interfaceTypeKey] = grainInterfaceData;
-                }
-
-                var implementation = new GrainClassData(grainTypeCode, grainClass, isGenericGrainClass, grainInterfaceData, placement, registrationStrategy);
+                var implementation = new GrainClassData(grainTypeCode, grainName, isGenericGrainClass, grainInterfaceData, placement, registrationStrategy);
                 if (!implementationIndex.ContainsKey(grainTypeCode))
                     implementationIndex.Add(grainTypeCode, implementation);
 
                 grainInterfaceData.AddImplementation(implementation, primaryImplementation);
                 if (primaryImplementation)
                 {
-                    primaryImplementations[grainInterface] = grainClass;
+                    primaryImplementations[grainInterfaceData.GrainInterface] = grainName;
                 }
                 else
                 {
-                    if (!primaryImplementations.ContainsKey(grainInterface))
-                        primaryImplementations.Add(grainInterface, grainClass);
+                    if (!primaryImplementations.ContainsKey(grainInterfaceData.GrainInterface))
+                        primaryImplementations.Add(grainInterfaceData.GrainInterface, grainName);
                 }
 
                 if (localTestMode)
                 {
+                    var assembly = grainTypeInfo.Assembly.CodeBase;
                     if (!loadedGrainAsemblies.Contains(assembly))
                         loadedGrainAsemblies.Add(assembly);
                 }
             }
         }
 
+        private GrainInterfaceData GetOrAddGrainInterfaceData(Type iface, bool isGenericGrainClass)
+        {
+            var interfaceId = GrainInterfaceUtils.GetGrainInterfaceId(iface);
+
+            // If already exist
+            GrainInterfaceData grainInterfaceData;
+            if (table.TryGetValue(interfaceId, out grainInterfaceData))
+                return grainInterfaceData;
+
+            // If not create new entry
+            var interfaceName = TypeUtils.GetRawClassName(TypeUtils.GetFullName(iface));
+            grainInterfaceData = new GrainInterfaceData(interfaceId, iface, interfaceName);
+            table[interfaceId] = grainInterfaceData;
+
+            // Add entry to mapping iface string -> data
+            var interfaceTypeKey = GetTypeKey(iface, isGenericGrainClass);
+            typeToInterfaceData[interfaceTypeKey] = grainInterfaceData;
+
+            // If we are adding a concrete implementation of a generic interface
+            // add also the latter to the map: GrainReference and InvokeMethodRequest 
+            // always use the id of the generic one
+            if (iface.IsConstructedGenericType)
+                GetOrAddGrainInterfaceData(iface.GetGenericTypeDefinition(), true);
```

Comment:
The meaning of the `isGenericGrainClass` parameter becomes a little strange here. This is correct, from what I can tell, but it's not obvious from the method signature that we should be passing true here. In effect, the `true` just informs the method that the "type key" should be a generic type definition (if it's a generic interface). An alternative would be to change the logic in `GetTypeKey` so that the if-statement reads:
`if (typeInfo.IsGenericTypeDefinition || (isGenericGrainClass && typeInfo.IsGenericType)`.

Does that make sense, or am I reasoning about it wrong?